### PR TITLE
[[AR]] - PAY-3126 and PAY-2876 - Spring boot upgrade and OIDC changes

### PR DESCRIPTION
**Pull request by aravindan89**
### JIRA link (if applicable) ###
1. https://tools.hmcts.net/jira/browse/PAY-3126
2. https://tools.hmcts.net/jira/browse/PAY-2876

### Change description ###
1. Spring boot upgrade
2. OIDC changes

**Does this PR introduce a breaking change?** (check one with "x")

```
[ x] Yes
[ ] No
```

### CODE REVIEW

File: api/build.gradle [hunk 1]
```diff
@@ -58,26 +58,45 @@ dependencies {
...
   compile group: 'com.microsoft.azure', name: 'azure-servicebus-spring-boot-starter', version: '2.1.6'
   compile group: 'org.springframework.boot', name: 'spring-boot-starter-cache'
   compile group: 'com.github.ben-manes.caffeine', name: 'caffeine', version: '2.7.0'
 
+    // Start of OpenId Connect. Few versions has been explicitly added as 5.2.2 due to lower spring boot version in payment-app.
+
+    compile group: 'io.jsonwebtoken', name: 'jjwt', version:'0.9.1'
+    compile group: 'uk.gov.hmcts.reform', name: 'service-auth-provider-client', version: '3.1.1'
+    compile group: 'uk.gov.hmcts.reform', name: 'idam-client', version: '1.3.1'
+    compile group: 'com.nimbusds', name: 'nimbus-jose-jwt', version: '7.9'
```
Comment (amitchidrewar1301) [inline api/build.gradle:85]: Please remove nimbus jar

File: api/src/functionalTest/java/uk/gov/hmcts/payment/functional/idam/IdamApi.java [hunk 2]
```diff
@@ -29,7 +29,7 @@ AuthenticateUserResponse authenticateUser(@Param("authorization") String authori
                                               @Param("client_id") String clientId,
                                               @Param("redirect_uri") String redirectUri);
```
Comment (amitchidrewar1301) [inline api/src/functionalTest/java/uk/gov/hmcts/payment/functional/idam/IdamApi.java:31]: Please refer to https://github.com/hmcts/ccpay-bulkscanning-app/commit/f6437a25140cb7cafd30b711741276ecc45195f2, need some code clean up

File: api/src/main/java/uk/gov/hmcts/PaymentApiApplication.java [hunk 3]
```diff
@@ -58,7 +68,10 @@ public CacheManager cacheManager() {
                     .build()),
             new CaffeineCache("sites", Caffeine.newBuilder()
                 .expireAfterWrite(48, TimeUnit.HOURS)
-                .build())
+                .build()),
+        new CaffeineCache("userInfoCache", Caffeine.newBuilder()
```
Comment (amitchidrewar1301) [inline api/src/main/java/uk/gov/hmcts/PaymentApiApplication.java:72]: Please remove this cache for userinfo and check if it works

File: api/src/main/java/uk/gov/hmcts/payment/api/configuration/SecurityUtils.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        } else {
+            authorities = authentication.getAuthorities();
+        }
+        return authorities.stream()
+            .map(GrantedAuthority::getAuthority);
+    }
+
+    @SuppressWarnings("unchecked")
+    public static List<GrantedAuthority> extractAuthorityFromClaims(Map<String, Object> claims) {
+        return ((List<String>) claims.get("roles"))
```
Comment (amitchidrewar1301) [inline api/src/main/java/uk/gov/hmcts/payment/api/configuration/SecurityUtils.java:62]: Please check null pointer for claims.get("roles")

File: api/src/main/java/uk/gov/hmcts/payment/api/configuration/converters/JwtGrantedAuthoritiesConverter.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        return authorities;
+    }
+
+    /**
+     * Method responsible to get stream of authorities based on claims
+     * @param roles
+     * @return
+     */
+    private List<GrantedAuthority> extractAuthorityFromClaims(List<String> roles) {
+        return roles.stream()
```
Comment (amitchidrewar1301) [inline api/src/main/java/uk/gov/hmcts/payment/api/configuration/converters/JwtGrantedAuthoritiesConverter.java:58]: null pointer check for roles

File: api/src/main/java/uk/gov/hmcts/payment/api/configuration/security/SpringSecurityConfiguration.java [hunk 6]
```diff
@@ -1,67 +1,113 @@
...
+import javax.servlet.http.HttpServletRequest;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Optional;
+import java.util.function.Function;
 
 import static org.springframework.security.config.http.SessionCreationPolicy.STATELESS;
 
-@EnableWebSecurity
+@EnableWebSecurity(debug = true)
```
Comment (amitchidrewar1301) [inline api/src/main/java/uk/gov/hmcts/payment/api/configuration/security/SpringSecurityConfiguration.java:32]: debug needs to be removed

File: api/src/main/java/uk/gov/hmcts/payment/api/configuration/security/SpringSecurityConfiguration.java [hunk 7]
```diff
@@ -1,67 +1,113 @@
...
 public class SpringSecurityConfiguration {
 
+    //private static final Logger LOG = LoggerFactory.getLogger(SpringSecurityConfiguration.class);
```
Comment (amitchidrewar1301) [inline api/src/main/java/uk/gov/hmcts/payment/api/configuration/security/SpringSecurityConfiguration.java:35]: This needs to be enabled, for logging the exceptions

File: api/src/main/java/uk/gov/hmcts/payment/api/configuration/security/authcheckerconfiguration/AuthCheckerConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        };
+    }
+
+    /**
+     * Bean to specify authorised roles for Fees&Pay
+     * @return
+     */
+    @Bean
+    public Function<HttpServletRequest, Collection<String>> authorizedRolesExtractor() {
+        return (any) -> Stream.of("payments", "citizen")
```
Comment (amitchidrewar1301) [inline api/src/main/java/uk/gov/hmcts/payment/api/configuration/security/authcheckerconfiguration/AuthCheckerConfiguration.java:83]: Please add all the authorised roles here i.e hasAnyAuthority("payments","pui-finance-manager","caseworker-cmc-solicitor", "caseworker-publiclaw-solicitor", "caseworker-probate-solicitor", "caseworker-financialremedy-solicitor", "caseworker-divorce-solicitor")

File: api/src/main/java/uk/gov/hmcts/payment/api/exception/UnauthorizedException.java [hunk 9]
```diff
@@ -0,0 +1,13 @@
+package uk.gov.hmcts.payment.api.exception;
+
+public class UnauthorizedException extends RuntimeException {
+
+    public UnauthorizedException(String message) {
+        super(message);
+    }
+
+    /*public UnauthorizedException(String message, Throwable cause) {
```
Comment (amitchidrewar1301) [inline api/src/main/java/uk/gov/hmcts/payment/api/exception/UnauthorizedException.java:9]: Remove this unused code

File: api/src/main/java/uk/gov/hmcts/PaymentApiApplication.java (hunk 3 above)
Comment (aravindan89) [reply api/src/main/java/uk/gov/hmcts/PaymentApiApplication.java:72]: Tried removing the **userInfoCache** but while authenticating the request we are not able to authorize the user.

File: api/src/main/java/uk/gov/hmcts/payment/api/configuration/SecurityUtils.java (hunk 4 above)
Comment (aravindan89) [reply api/src/main/java/uk/gov/hmcts/payment/api/configuration/SecurityUtils.java:62]: Changes done.

File: api/src/main/java/uk/gov/hmcts/payment/api/configuration/converters/JwtGrantedAuthoritiesConverter.java (hunk 5 above)
Comment (aravindan89) [reply api/src/main/java/uk/gov/hmcts/payment/api/configuration/converters/JwtGrantedAuthoritiesConverter.java:58]: Changes done.

File: api/src/main/java/uk/gov/hmcts/payment/api/configuration/security/SpringSecurityConfiguration.java (hunk 6 above)
Comment (aravindan89) [reply api/src/main/java/uk/gov/hmcts/payment/api/configuration/security/SpringSecurityConfiguration.java:32]: Point noted. Will remove before merging to master.

File: api/src/main/java/uk/gov/hmcts/payment/api/configuration/security/SpringSecurityConfiguration.java (hunk 7 above)
Comment (aravindan89) [reply api/src/main/java/uk/gov/hmcts/payment/api/configuration/security/SpringSecurityConfiguration.java:35]: Changes done.

File: api/src/main/java/uk/gov/hmcts/payment/api/configuration/security/authcheckerconfiguration/AuthCheckerConfiguration.java (hunk 8 above)
Comment (aravindan89) [reply api/src/main/java/uk/gov/hmcts/payment/api/configuration/security/authcheckerconfiguration/AuthCheckerConfiguration.java:83]: Changes done.

File: api/src/main/java/uk/gov/hmcts/payment/api/exception/UnauthorizedException.java (hunk 9 above)
Comment (aravindan89) [reply api/src/main/java/uk/gov/hmcts/payment/api/exception/UnauthorizedException.java:9]: Changes done.

File: api/src/functionalTest/java/uk/gov/hmcts/payment/functional/idam/IdamApi.java (hunk 2 above)
Comment (aravindan89) [reply api/src/functionalTest/java/uk/gov/hmcts/payment/functional/idam/IdamApi.java]: Changes done.

File: api/src/main/java/uk/gov/hmcts/payment/api/configuration/converters/JwtGrantedAuthoritiesConverter.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+        return authorities;
+    }
+
+    /**
+     * Method responsible to get stream of authorities based on claims
+     * @param roles
+     * @return
+     */
+    private List<GrantedAuthority> extractAuthorityFromClaims(List<String> roles) {
+        //
```
[the remark this answers is not in the document]
Comment (aravindan89) [reply api/src/main/java/uk/gov/hmcts/payment/api/configuration/converters/JwtGrantedAuthoritiesConverter.java:60]: Removed.

File: api/src/main/resources/application.properties [hunk 11]
```diff
@@ -41,7 +41,19 @@ gov.pay.operational_services=${GOV_PAY_OPERATIONAL_SERVICES:#{ccd_gw,api_gw}}
 
 # S2S trusted services
 
+# Required for the ServiceAuthorisationApi class in service-auth-provider-java-client library
+idam.s2s-auth.totp_secret=${OIDC_S2S_SECRET}
+idam.s2s-auth.microservice=${OIDC_S2S_MICROSERVICE_NAME}
+idam.s2s-authorised.services=cmc,cmc_claim_store,probate_frontend,divorce_frontend,ccd_gw,bar_api,api_gw,finrem_payment_service,ccpay_bubble,jui_webapp,xui_webapp,fpl_case_service
+idam.api.url=${IDAM_API_URL:https://idam-api.aat.platform.hmcts.net}
+idam.s2s-auth.url=${AUTH_PROVIDER_SERVICE_CLIENT_BASEURL}
+logging.level.org.springframework.security=DEBUG
```
[the remark this answers is not in the document]
Comment (aravindan89) [reply api/src/main/resources/application.properties:50]: Yes, we will remove it while moving to master.

File: api/src/functionalTest/java/uk/gov/hmcts/payment/functional/idam/IdamService.java [hunk 12]
```diff
@@ -50,7 +47,13 @@ public IdamService(TestConfigProperties testConfig) {
     public User createUserWith(String userGroup, String... roles) {
         String email = nextUserEmail();
         CreateUserRequest userRequest = userRequest(email, userGroup, roles);
-        idamApi.createUser(userRequest);
+        LOG.info("idamApi : " + idamApi.toString());
```
[the remark this answers is not in the document]
Comment (aravindan89) [reply api/src/functionalTest/java/uk/gov/hmcts/payment/functional/idam/IdamService.java:50]: These logs added for debug purpose we will remove it while moving to master.

File: api/src/functionalTest/java/uk/gov/hmcts/payment/functional/idam/IdamService.java [hunk 13]
```diff
@@ -64,30 +67,37 @@ public String authenticateUser(String username, String password) {
...
-        TokenExchangeResponse tokenExchangeResponse = idamApi.exchangeCode(
-            authenticateUserResponse.getCode(),
-            AUTHORIZATION_CODE,
-            testConfig.getOauth2().getClientId(),
-            testConfig.getOauth2().getClientSecret(),
-            testConfig.getOauth2().getRedirectUrl()
-        );
-
-        return BEARER + tokenExchangeResponse.getAccessToken();
+        LOG.info("username : " + username);
```
[the remark this answers is not in the document]
Comment (aravindan89) [reply api/src/functionalTest/java/uk/gov/hmcts/payment/functional/idam/IdamService.java:70]: These logs added for debug purpose we will remove it while moving to master.

File: api/src/test/java/uk/gov/hmcts/payment/api/componenttests/PaymentAllocationControllerTest.java [hunk 14]
```diff
@@ -38,36 +46,46 @@
...
-    private RestActions restActions;
+    @MockBean
+    private ClientRegistrationRepository clientRegistrationRepository;
+
+    @MockBean
+    private JwtDecoder jwtDecoder;
 
     @Autowired
-    private ObjectMapper objectMapper;
+    ServiceAuthFilter serviceAuthFilter;
```
[the remark this answers is not in the document]
Comment (aravindan89) [reply api/src/test/java/uk/gov/hmcts/payment/api/componenttests/PaymentAllocationControllerTest.java:79]: Changed to private.

File: api/src/test/java/uk/gov/hmcts/payment/api/componenttests/PaymentAllocationControllerTest.java [hunk 15]
```diff
@@ -38,36 +46,46 @@
...
 
     @Autowired
-    private SiteService<Site, String> siteServiceMock;
+    ServicePaymentFilter servicePaymentFilter;
```
[the remark this answers is not in the document]
Comment (aravindan89) [reply api/src/test/java/uk/gov/hmcts/payment/api/componenttests/PaymentAllocationControllerTest.java:82]: Changed to private.

**Pull request closed without merging**